Issues with type hints for `urllib.parse.urldefrag`

- [ ] PY3: `reveal_type(urldefrag(...)[1])` should retain the type of the input instead of `Any`
- [x] PY2: `reveal_type(urldefrag(...)[1])` should not be fixed as `str`; should be `unicode` when a unicode input is provided
- [x] PY2: type for `urldefrag` should be updated to allow unicode input

``` python
from six.moves import urllib
from typing_extensions import TYPE_CHECKING

if not TYPE_CHECKING:
    def reveal_type(ignored):
        pass

_, frg = urllib.parse.urldefrag("https://github.com/python/typeshed/#typeshed")
print(type(frg))
reveal_type(frg)

_, frg2 = urllib.parse.urldefrag(u"https://github.com/python/typeshed/#typeshed")
print(type(frg2))
reveal_type(frg2)

_, frg3 = urllib.parse.urldefrag(b"https://github.com/python/typeshed/#typeshed")
print(type(frg3))
reveal_type(frg3)
```


```
$ mypy  typing-defrag-test.py                                                                                          
typing-defrag-test.py:11: error: Revealed type is 'Any'
typing-defrag-test.py:15: error: Revealed type is 'Any'
typing-defrag-test.py:19: error: Revealed type is 'Any'

$ mypy --py2  typing-defrag-test.py                                                                                    
typing-defrag-test.py:11: error: Revealed type is 'builtins.str'
typing-defrag-test.py:13: error: Argument 1 to "urldefrag" has incompatible type "unicode"; expected "str"
typing-defrag-test.py:15: error: Revealed type is 'builtins.str'
typing-defrag-test.py:19: error: Revealed type is 'builtins.str'
```
```
$ python3 typing-defrag-test.py                                                                                         
<class 'str'>
<class 'str'>
<class 'bytes'>

$ python2 typing-defrag-test.py                                                                                        
<type 'str'>
<type 'unicode'>
<type 'str'>
```
```
$ mypy --version
mypy 0.630
$ python -V
Python 2.7.15+
$ python3 -V
Python 3.6.6+
```

--- a/stdlib/urllib/parse.pyi
+++ b/stdlib/urllib/parse.pyi
@@ -66,7 +66,7 @@ class _NetlocResultMixinBase(Generic[AnyStr]):
 class _NetlocResultMixinStr(_NetlocResultMixinBase[str], _ResultMixinStr): ...
 class _NetlocResultMixinBytes(_NetlocResultMixinBase[bytes], _ResultMixinBytes): ...
 
-class _DefragResultBase(tuple[Any, ...], Generic[AnyStr]):
+class _DefragResultBase(tuple[AnyStr, ...], Generic[AnyStr]):
     if sys.version_info >= (3, 10):
         __match_args__ = ("url", "fragment")
     @property

--- a/stdlib/urllib/parse.pyi
+++ b/stdlib/urllib/parse.pyi
@@ -66,6 +66,8 @@ class _NetlocResultMixinBase(Generic[AnyStr]):
 class _NetlocResultMixinStr(_NetlocResultMixinBase[str], _ResultMixinStr): ...
 class _NetlocResultMixinBytes(_NetlocResultMixinBase[bytes], _ResultMixinBytes): ...
 
+# Ideally this would be a generic fixed-length tuple,
+# but mypy doesn't support that yet: https://github.com/python/mypy/issues/685#issuecomment-992014179
 class _DefragResultBase(tuple[AnyStr, ...], Generic[AnyStr]):
     if sys.version_info >= (3, 10):
         __match_args__ = ("url", "fragment")

